src/watershed.jl

```diff
@@ -39,22 +51,61 @@ function watershed(img::AbstractArray{T, N}, markers::AbstractArray{S,N}) where
             for j in CartesianIndices(_colon(max(Istart,i-one(i)), min(i+one(i),Iend)))
                 if segments[j] == 0
                     segments[j] = markers[i]
-                    enqueue!(pq, j, PixelKey(img[i], time_step))
+                    enqueue!(pq, j, PixelKey(img[i], time_step, j))
                     time_step += 1
                 end
             end
         end
     end
 
     while !isempty(pq)
-        current = dequeue!(pq)
-        segments_current = segments[current]
-        img_current = img[current]
-        for j in CartesianIndices(_colon(max(Istart,current-one(current)), min(current+one(current),Iend)))
+        curr_idx, curr_elem = dequeue_pair!(pq)
+        segments_current = segments[curr_idx]
+
+        # If we're using the compact algorithm, we need assign grouping for a given location
+        # when it comes off the queue since we could have found a better suited watershed later.
+        if compact
+            if segments_current > 0 && curr_idx != curr_elem.source
+                # this is a non-marker location that we've already assigned
+                continue
+            end
+            # group this location with its watershed
+            segments[curr_idx] = segments[curr_elem.source]
+        end
+
+        img_current = img[curr_idx]
+        for j in CartesianIndices(_colon(max(Istart,curr_idx-one(curr_idx)), min(curr_idx+one(curr_idx),Iend)))
+
+            # if this location is false in the mask, we skip it
+            (!mask[j]) && continue
+            # only continue if this is a position that we haven't assigned yet
             if segments[j] == 0
-                segments[j] = segments_current
-                enqueue!(pq, j, PixelKey(img_current, time_step))
-                time_step += 1
+                # if we're doing a simple watershed, we can go ahead and set the final grouping for a new
+                # ungrouped position the moment we first encounter it
+                if !compact
+                    segments[j] = segments_current
+                    new_value = img_current
+                else
+                    # in the compact algorithm case, we don't set the grouping at push-time and calculate
+                    # a weighted value based on the
```

Comment:
this seems incomplete or confusing. Do we NOT set the grouping at push-time and also NOT calculate a weighted value...? Or is calculating the weighted value something we do INSTEAD?